packages/salesforcedx-vscode-apex/src/embeddedSoql/soqlCompletion.ts

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {
+  commands,
+  CompletionItem,
+  CompletionList,
+  EndOfLine,
+  Position,
+  TextDocument,
+  Uri,
+  workspace
+} from 'vscode';
+
+import { Middleware } from 'vscode-languageclient/lib/main';
+import ProtocolCompletionItem from 'vscode-languageclient/lib/protocolCompletionItem';
+
+const SOQL_SPECIAL_COMPLETION_ITEM_LABEL = '_SOQL_';
+
+const virtualDocumentContents = new Map<string, string>();
+
+workspace.registerTextDocumentContentProvider('embedded-soql', {
+  provideTextDocumentContent: uri => {
+    const originalUri = uri.path.replace(/^\//, '').replace(/.soql$/, '');
+
+    const decodedUri = decodeURIComponent(originalUri);
+    return virtualDocumentContents.get(decodedUri);
+  }
+});
+
+function insideSOQLBlock(
+  apexItems: ProtocolCompletionItem[]
+): { queryText: string; location: any } | undefined {
+  const soqlItem = apexItems.find(
+    i => i.label === SOQL_SPECIAL_COMPLETION_ITEM_LABEL
+  );
+  return soqlItem
+    ? { queryText: soqlItem.detail as string, location: soqlItem.data }
+    : undefined;
+}
+function insideApexBindingExpression(
+  document: TextDocument,
+  soqlQuery: string,
+  position: Position
+): boolean {
+  // Simple heuristic to detect when cursor is on a binding expression
+  // (which might have been missed by Apex LSP)
+  const rangeAtCursor = document.getWordRangeAtPosition(
+    position,
+    /[:(_\.\w)]+/
+  );
+  const wordAtCursor = rangeAtCursor
+    ? document.getText(rangeAtCursor)
+    : undefined;
+
+  return !!wordAtCursor && wordAtCursor.startsWith(':');
+}
```

Comment:
What makes the VS Code context more likely to catch this scenario of whether or not the cursor is in a binding expression? Is it not possible for the extra check to be in the language server?

Reply:
I agree that this belongs to the server side.... however, this is a "hack", which is much simpler to implement here. The real solution is to fix the start/end indexes bug on the language server side.

Reply:
Gotcha. Proper solution aside, I guess what makes the "hack" easier to do here than have the "hack" in the language server?